--- opt/krona/updateTaxonomy.sh	2016-04-20 17:45:37.000000000 -0400
+++ /tmp/.psub.mmObSlLDI8	2016-04-27 02:10:19.894387715 -0400
@@ -56,9 +56,9 @@
 	
 	if [ $local ]
 	then
-		if [ ! -e $zipped ]
+		if [ ! -e $zipped -a ! -e $unzipped ]
 		then
-			die "Could not find $taxonomyPath/$zipped.  Was it transfered?"
+			die "Could not find $taxonomyPath/$unzipped[.gz].  Was it transfered?"
 		fi
 	else
         if [ -e $timestamp ]
@@ -112,16 +112,19 @@
 	fi
 fi
 
-cd $taxonomyPath;
-
 if [ "$?" != "0" ]
 then
 	die "Could not enter '$oldPath/taxonomy'. Did you run install.pl?"
 fi
 
+pushd "$taxonomyPath" > /dev/null
 update gi_taxid_nucl.dmp gi_taxid.dat "GI to taxID dump (nucleotide)"
 update gi_taxid_prot.dmp gi_taxid.dat "GI to taxID dump (protein)"
+if [ -z $local -o ! -e names.dmp ]
+then
 update taxdump.tar taxonomy.tab 'Taxonomy dump'
+fi
+popd > /dev/null
 
 if [ -e taxdump.tar ]
 then
@@ -163,4 +166,3 @@
 echo
 echo ">>>>> Finished."
 echo
-
